vizro-core/tests/unit/vizro/models/_components/test_container.py

```diff
@@ -0,0 +1,57 @@
+"""Unit tests for vizro.models.Container."""
+import pytest
+from asserts import assert_component_equal
+from dash import html
+
+try:
+    from pydantic.v1 import ValidationError
+except ImportError:  # pragma: no cov
+    from pydantic import ValidationError
+
+import vizro.models as vm
+
+
+class TestContainerInstantiation:
+    """Tests model instantiation and the validators run at that time."""
+
+    def test_create_container_mandatory_only(self):
+        container = vm.Container(title="Title", components=[vm.Button(), vm.Button()])
+        assert isinstance(container.components[0], vm.Button) and isinstance(container.components[1], vm.Button)
+        assert container.layout.grid == [[0], [1]]
+        assert container.title == "Title"
+
+    def test_create_container_mandatory_and_optional(self):
+        container = vm.Container(
+            title="Title",
+            components=[vm.Button(), vm.Button()],
+            id="my-id",
+            layout=vm.Layout(grid=[[0, 1]]),
+        )
+        assert isinstance(container.components[0], vm.Button) and isinstance(container.components[1], vm.Button)
+        assert container.layout.grid == [[0, 1]]
+        assert container.title == "Title"
+        assert container.id == "my-id"
+
+    def test_mandatory_title_missing(self):
+        with pytest.raises(ValidationError, match="field required"):
+            vm.Container(components=[vm.Button()])
+
+    def test_mandatory_components_missing(self):
+        with pytest.raises(ValidationError, match="field required"):
+            vm.Container(title="Title")
+
+
+class TestContainerBuildMethod:
+    def test_container_build(self):
+        result = vm.Container(id="container", title="Title", components=[vm.Button()]).build()
+        assert_component_equal(
+            result, html.Div(className="page-component-container", id="container"), keys_to_strip={"children"}
+        )
+        assert_component_equal(
+            result.children,
+            [html.H3("Title"), html.Div([html.Div(vm.Button().build())])],
+            # TODO: Previously I've ignored id, className and style as I had the helper function tested in
+            # test_layout.py. With the new changes it seems like we shouldn't strip them now?
+            # Shall we then need to duplicate for Page and Form as well?
+            keys_to_strip={"id", "className", "style"},
+        )
```

Comment:
After much thought, this is what I think we should go for:

```suggestion
    def test_container_build(self):
        result = vm.Container(
            id="container", title="Title", components=[vm.Button()], layout=vm.Layout(id="layout_id", grid=[[0]])
        ).build()
        assert_component_equal(
            result, html.Div(className="page-component-container", id="container"), keys_to_strip={"children"}
        )
        assert_component_equal(result.children, [html.H3(), html.Div()], keys_to_strip=STRIP_ALL)
        # We still want to test the exact H3 produced in Container.build:
        assert_component_equal(result.children[0], html.H3("Title"))
        # And also that a button has been inserted in the right place:
        assert_component_equal(result["layout_id_0"].children.children, dbc.Button(), keys_to_strip=STRIP_ALL)
```

Note:
* it's best to avoid explicit calls to e.g. `vm.Button.build()` in these tests - instead better to just look at what you expect the result of that call to be (`dbc.Button`)
* we want to test specifically for things done by `Container.build`, e.g. the `H3("TItle")` and correct use of the `"layout_id_0"` index, but just generically for things that it's not responsible for e.g. `dbc.Button()`
* the above is not very nice tbh: `assert_component_equal` does not work very well in this case. But generally you can achieve the degree of precision that you need by using it repeatedly

Reply:
Yes, that makes a lot of sense! I tried to squeeze everything into one test but then struggled when I wanted to explain your point 2. Repeating makes sense to me here 👍